src/pudl/io_managers.py

```diff
@@ -729,6 +729,36 @@ def filter_for_freshest_data(
 
         return deduped
 
+    @staticmethod
+    def refine_report_year(df: pd.DataFrame) -> pd.DataFrame:
+        """Set a fact's report year by its actual dates.
+
+        Sometimes a fact belongs to a context which has no ReportYear
+        associated with it; other times there are multiple ReportYears
+        associated with a single filing. In these cases the report year of a
+        specific fact may be associated with the other years in the filing.
+
+        In many cases we can infer the actual report year from the fact's
+        associated time period - either duration or instant.
+        """
+        is_duration = len({"start_date", "end_date"} - set(df.columns)) == 0
+        is_instant = "date" in df.columns
+
+        def get_year(s: pd.Series) -> pd.Series:
+            return pd.to_datetime(s).apply(lambda x: x.year)
+
+        if is_duration:
+            start_years = get_year(df.start_date)
+            end_years = get_year(df.end_date)
+            replace_mask = (start_years == end_years) & start_years.notna()
```

Comment:
If `start_year != end_year` should we infer the report year as `end_year`? Or just keep the original report year?

Reply:
Talked with @zaneselvans - we'll actually just assert a bunch of things about this data and fail when the data doesn't match our assumptions.